web-common/src/features/canvas/stores/canvas-entity.ts

```diff
@@ -177,39 +181,106 @@ export class CanvasEntity {
       }
     });
 
-    // TODO: merge more stores once we add support for defaults for those.
-    this.defaultUrlParamsStore = this.timeControls.defaultUrlParamsStore;
-
     this.theme = createResolvedThemeStore(
       this.themeName,
       this.specStore,
       this.instanceId,
     );
   }
 
-  onUrlParamsChange = (urlParams: URLSearchParams) => {
+  onUrlParamsChange = async (
+    urlParams: URLSearchParams,
+    builderContext?: boolean,
+  ) => {
+    if (builderContext) {
+      const redirected = await CanvasEntity.handleCanvasRedirect({
+        canvasName: this.name,
+        searchParams: urlParams,
+        pathname: window.location.pathname,
+        builderContext: true,
+      });
+
+      if (redirected) return;
+    }
+
+    this.searchParams.set(urlParams);
     this.themeName.set(urlParams.get("theme") ?? undefined);
+    this.saveSnapshot(urlParams.toString());
   };
 
   // Not currently being used
   unsubscribe = () => {
     // this.unsubscriber();
   };
 
-  saveSnapshot = (filterState: string) => {
-    this.lastVisitedState.set(filterState);
-  };
+  static handleCanvasRedirect = async ({
+    canvasName,
+    searchParams,
+    pathname,
+    projectId,
+    builderContext,
+  }: {
+    canvasName: string;
+    searchParams: URLSearchParams;
+    pathname: string;
+    projectId?: string;
+    builderContext?: true;
+  }) => {
+    // If there are no URL params, check for last visited state or home bookmark
+    if (searchParams.size === 0) {
+      const snapshotSearchParams = lastVisitedState.get(canvasName);
+
+      if (snapshotSearchParams) {
+        if (builderContext) {
+          await goto(`?${snapshotSearchParams}`, { replaceState: true });
+          return true;
+        } else {
+          throw redirect(307, `?${snapshotSearchParams}`);
+        }
+      }
 
-  restoreSnapshot = async (initState: string | undefined) => {
-    const stateToRestore = get(this.lastVisitedState) ?? initState;
+      if (projectId && !builderContext) {
+        let homeBookmarkUrlSearch: string | undefined = undefined;
+        try {
+          const { adminServiceListBookmarks } = await import(
+            "@rilldata/web-admin/client"
+          );
 
-    if (stateToRestore) {
-      await goto(`?${stateToRestore}`, {
-        replaceState: true,
-      });
+          const response = await adminServiceListBookmarks({
```

Comment:
Any reason to not use query client's fetch query? This is leading to double call to bookmarks API